app/controllers/collections_controller.rb

```diff
@@ -121,6 +121,93 @@ def revoke_access
     redirect_to licences_path
   end
 
+  def add_items_to_collection
```

Comment:
This is *really* long method, which is what some people call a [code smell](https://en.wikipedia.org/wiki/Code_smell). Even though it is perfectly functional, it is probably doing a lot of work and a lot of different things, which makes it harder for other people to understand.

In this case, there is lots of validation code interspersed with processing code, you could probably move all the code to a `validate_add_items_request` type method. That method would itself just call different helper methods `validate_item`, `validate_document`, etc. which throw Exceptions with the appropriate message and response code.

The main block would then look something like:

```
def add_items_to_collection
  begin
    validate_add_items_request(params)
    ...
  rescue SomeException
    respond_with_error(e.message, e.response_code)
  end
end
```